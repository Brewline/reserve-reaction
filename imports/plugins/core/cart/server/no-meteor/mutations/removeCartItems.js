import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import hashLoginToken from "/imports/node-app/core/util/hashLoginToken";

const inputSchema = new SimpleSchema({
  "cartId": String,
  "cartItemIds": {
    type: Array,
    minCount: 1
  },
  "cartItemIds.$": String,
  "token": {
    type: String,
    optional: true
  }
});

/**
 * @method removeCartItems
 * @summary Removes one or more items from a cart
 * @param {Object} context -  an object containing the per-request state
 * @param {Object} input - Necessary input
 * @param {String} input.cartId - The ID of the cart in which all of the items exist
 * @param {String[]} input.cartItemIds - Array of cart item IDs to remove
 * @param {String} input.token - The token if the cart is an anonymous cart
 * @return {Promise<Object>} An object containing the updated cart in a `cart` property
 */
export default async function removeCartItems(context, input) {
  inputSchema.validate(input || {});

  const { accountId, appEvents, collections, userId } = context;
  const { Cart } = collections;
  const { cartId, cartItemIds, token } = input;

  const selector = { _id: cartId };
  if (token) {
    selector.anonymousAccessToken = hashLoginToken(token);
  } else if (accountId) {
    selector.accountId = accountId;
  } else {
    throw new ReactionError("invalid-param", "A token is required when updating an anonymous cart");
  }

  const { modifiedCount } = await Cart.updateOne(selector, {
    $pull: {
      items: {
        _id: { $in: cartItemIds }
      }
    }
  });
  if (modifiedCount === 0) throw new ReactionError("not-found", "Cart not found or provided items are not in the cart");

  const cart = await Cart.findOne(selector);
  if (!cart) throw new ReactionError("not-found", "Cart not found");

  await appEvents.emit("afterCartUpdate", {
    cart,
    updatedBy: userId
  });

  return { cart };
}
