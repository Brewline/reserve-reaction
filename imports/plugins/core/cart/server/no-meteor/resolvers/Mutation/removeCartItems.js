import { decodeCartItemOpaqueId, decodeCartOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/cart";

/**
 * @name "Mutation.removeCartItems"
 * @method
 * @memberof Cart/GraphQL
 * @summary resolver for the removeCartItems GraphQL mutation
 * @param {Object} parentResult - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} args.input.cartId - The ID of the cart in which all of the items exist
 * @param {String[]} args.input.cartItemIds - Array of item IDs to update
 * @param {String} args.input.token - The token if the cart is an anonymous cart
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Object>} RemoveCartItemsPayload
 */
export default async function removeCartItems(parentResult, { input }, context) {
  const { cartId: opaqueCartId, clientMutationId = null, cartItemIds: opaqueCartItemIds, token } = input;

  const cartId = decodeCartOpaqueId(opaqueCartId);
  const cartItemIds = opaqueCartItemIds.map(decodeCartItemOpaqueId);

  const { cart } = await context.mutations.removeCartItems(context, {
    cartId,
    cartItemIds,
    token
  });

  return {
    cart,
    clientMutationId
  };
}
