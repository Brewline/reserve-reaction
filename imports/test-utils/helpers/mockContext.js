const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
