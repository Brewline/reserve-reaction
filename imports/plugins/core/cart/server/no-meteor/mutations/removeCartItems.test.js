import mockContext from "/imports/test-utils/helpers/mockContext";
import removeCartItems from "./removeCartItems";

const dbCart = {
  _id: "cartId",
  items: []
};

const cartItemIds = ["cartItemId1", "cartItemId2"];

test("removes multiple items from account cart", async () => {
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(dbCart));

  const result = await removeCartItems(mockContext, { cartId: "cartId", cartItemIds });

  expect(mockContext.collections.Cart.updateOne).toHaveBeenCalledWith({
    _id: "cartId",
    accountId: "FAKE_ACCOUNT_ID"
  }, {
    $pull: {
      items: {
        _id: {
          $in: cartItemIds
        }
      }
    }
  });

  expect(mockContext.collections.Cart.findOne).toHaveBeenCalledWith({
    _id: "cartId",
    accountId: "FAKE_ACCOUNT_ID"
  });

  expect(result).toEqual({ cart: dbCart });
});

test("updates the quantity of multiple items in anonymous cart", async () => {
  const hashedToken = "+YED6SF/CZIIVp0pXBsnbxghNIY2wmjIVLsqCG4AN80=";

  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(dbCart));

  const cachedAccountId = mockContext.accountId;
  mockContext.accountId = null;
  const result = await removeCartItems(mockContext, {
    cartId: "cartId",
    cartItemIds,
    token: "TOKEN"
  });
  mockContext.accountId = cachedAccountId;

  expect(mockContext.collections.Cart.updateOne).toHaveBeenCalledWith({
    _id: "cartId",
    anonymousAccessToken: hashedToken
  }, {
    $pull: {
      items: {
        _id: {
          $in: cartItemIds
        }
      }
    }
  });

  expect(mockContext.collections.Cart.findOne).toHaveBeenCalledWith({
    _id: "cartId",
    anonymousAccessToken: hashedToken
  });

  expect(result).toEqual({ cart: dbCart });
});

test("throws when no account and no token passed", async () => {
  const cachedAccountId = mockContext.accountId;
  mockContext.accountId = null;

  await expect(removeCartItems(mockContext, {
    cartId: "cartId",
    cartItemIds
  })).rejects.toThrowErrorMatchingSnapshot();

  mockContext.accountId = cachedAccountId;
});
