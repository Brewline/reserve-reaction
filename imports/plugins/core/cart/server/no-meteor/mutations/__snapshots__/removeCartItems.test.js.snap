// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws when no account and no token passed 1`] = `"A token is required when updating an anonymous cart"`;
